### PR TITLE
Introduce V3 to fix missing snapshot manifests

### PR DESCRIPTION
**Pull request by charleso**
/cc @raronson @nhibberd

### CODE REVIEW

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV0.scala [hunk 1]
```diff
@@ -73,10 +77,11 @@ object UpdateV0 {
 
     val key = Key(KeyName.unsafe(".metadata.json"))
 
-    def read(repository: Repository, s: Key): RIO[Option[SnapshotManifest]] = for {
-      exists <- repository.store.exists(s / SnapshotMetadataV1.key)
+    def read(repository: Repository, snapshotId: SnapshotId): RIO[Option[SnapshotManifest]] = for {
+      path   <- RIO.ok(Repository.snapshot(snapshotId) / key)
+      exists <- repository.store.exists(path)
       sm <- if (exists) for {
-        json <- repository.store.utf8.read(s / SnapshotMetadataV1.key)
```
Comment (charleso) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV0.scala]: My bad, this never worked. :(

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV0.scala [hunk 2]
```diff
@@ -19,17 +19,20 @@ object UpdateV0 {
...
+  def updateFactsets: RepositoryTIO[Unit] = RepositoryT.fromRIO(repository =>
+    Metadata.findFactsets(repository).flatMap(updateFactsetsIds(repository, _))
+  )
+
+  def updateFactsetsIds(repository: Repository, factsets: List[FactsetId]): RIO[Unit] = for {
     _ <- factsets.traverseU(f => for {
-      partitions <- Partitions.scrapeFromFactset(repository, f).map(p => (!p.isEmpty).option(p))
-      _    <- partitions.traverseU(ps => FactsetManifest.io(repository, f).write(FactsetManifest.create(f, FactsetFormat.V2, ps)))
+      partitions <- Partitions.scrapeFromFactset(repository, f)
+      _    <- FactsetManifest.io(repository, f).write(FactsetManifest.create(f, FactsetFormat.V2, partitions))
```
Comment (raronson) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV0.scala]: Should this be `V2` or `V3`?

Comment (charleso) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV0.scala]: > Should this be V2 or V3?

This is `FactsetFormat` which is different. :)

Comment (raronson) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV0.scala]: ah yep, sorry

**Comment by raronson** [comment]
LGTM

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV2.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import scalaz._, Scalaz._
+
+object UpdateV2 {
+
+  def update: RepositoryTIO[Unit] =
+    for {
+      // The V0 code has been fixed, and we just need to re-run
+      _ <- UpdateV0.updateSnapshots
+      // Even empty factsets should have a manifest
+      _ <- updateEmptyFactsts
```
Comment (markhibberd) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV2.scala]: This should not happen, if you need to do this, it is covering a far worse bug in the factset read code. This needs to be changed back.

Comment (raronson) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/update/UpdateV2.scala]: Im just curious - It is possible that empty factsets have been created and added to a store (it has happened in an existing repository). From my understanding, the previous updates did not add a manifest to empty factsets (even those within a store). Given this, how do you know if a factset should be skipped or hard fail in the factset read code? The contents of a factset dir might be accidentally removed and im not sure of a way to distinguish the difference between a valid empty factset (within a store) and one which has its contents removed.

**Pull request merged**